src/Microsoft.ML.TimeSeries/STL/Loess.cs

```diff
@@ -0,0 +1,206 @@
+﻿using System;
+using System.Collections.Generic;
+using Microsoft.ML.Runtime;
+
+namespace Microsoft.ML.TimeSeries
+{
+    /// <summary>
+    /// Loess is short for Robust Locally Weighted Regression and Smoothing Scatterplots.
+    /// </summary>
+    internal class Loess
+    {
+
+        private const double NumericalThreshold = 1.0e-10;
+
+        /// <summary>
+        /// the ratio to determine the local region
+        /// </summary>
+        private readonly int _r;
+
+        private readonly bool _isTemporal;
+
+        /// <summary>
+        /// key is the index of the given point, value is the corresponding neighbors of the given point.
+        /// </summary>
+        private readonly Dictionary<int, LocalRegression> _neighbors;
+
+        private IReadOnlyList<double> _x;
+        private IReadOnlyList<double> _y;
+        private readonly int _length;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Loess"/> class.
+        /// constructing the least square algorithm
+        /// </summary>
+        /// <param name="xValues">the corresponding x-axis value</param>
+        /// <param name="yValues">the corresponding y-axis value</param>
+        /// <param name="isTemporal">if the regression is considered to take temporal information into account. in general, this is true if we are regressing a time series, and false if we are regressing scatter plot data</param>
+        public Loess(IReadOnlyList<double> xValues, IReadOnlyList<double> yValues, bool isTemporal)
+        {
+            Contracts.CheckValue(xValues, nameof(xValues));
+            Contracts.CheckValue(yValues, nameof(yValues));
+
+            if (xValues.Count < BasicParameters.MinTimeSeriesLength || yValues.Count < BasicParameters.MinTimeSeriesLength)
+                throw Contracts.Except("input data structure cannot be 0-length: lowess");
+
+            if (xValues.Count != yValues.Count)
+                throw Contracts.Except("the x-axis length should be equal to y-axis length!: lowess");
+
+            _neighbors = new Dictionary<int, LocalRegression>();
+
+            _length = xValues.Count;
+            _isTemporal = isTemporal;
+
+            _r = (int)(_length * LoessConfiguration.F);
+
+            // r cannot be equal to length.
+            if (_r >= _length)
+                _r = _length - 1;
+            else if (_r < LoessConfiguration.MinimumNeighborCount) // the neighbors should be at least 2, or the matrix operations would encounter issues.
+                _r = LoessConfiguration.MinimumNeighborCount;
+
+            // DEBUG
+            // control the performance
+            if (_r >= LoessConfiguration.MaximumNeighborCount)
+            {
+                _r = LoessConfiguration.MaximumNeighborCount;
+            }
```

Comment:
Is there a reason to include this in here but not in the other Loess constructor?

Reply:
The effect of the optimization is under investigation, so remove it temporarily